Add xchacha20 and XChacha20Poly1305

kdenhartog opened the pull request:
XChacha20 module appears to be good and has passing tests. Would you mind giving it a review? 

Additionally, I'm working through the XChacha20Poly1305 implementation based on [draft-arciszewski-xchacha-03](https://tools.ietf.org/html/draft-arciszewski-xchacha-03#section-2), but have encountered some problems based on the suggested method of implementation in the RFC. Specifically in section 2, it says to generate the subkey using the Hchacha method and then to generate the nonce, which I've tried. However, it doesn't specify which of the APIs should be used which is what I'm working through now.

I'd appreciate some feedback on this work in order to get the test passing which comes from the RFC as well as any feedback about code style would be appreciated as well.

Thanks

kdenhartog commented:
If it would be easier for you, I can split this into two individual PRs. One for the XChacha20 cipher work, and then a separate one for the XChacha20Poly1305 work which is still in progress.

kdenhartog commented:
Got the tests passing for this PR and cleaned up the commit log so it's easier to review based on the specific packages.

dchest commented:
Thank you! Will review on Thursday.

dchest commented on packages/xchacha20poly1305/xchacha20poly1305.ts:
```diff
@@ -0,0 +1,169 @@
+   * draft-irtf-cfrg-xchacha defines a 24 byte nonce (192 bits) which
+   * then uses the first 16 bytes of the nonce and the secret key with
+   * Hchacha to generate an initial subkey. The last 8 bytes of the nonce
+   * are then prefixed with 4 null bytes and then provided with the subkey
```
"4 zero bytes"

kdenhartog replied:
Fixed for all of the times where it occurred.

dchest commented on packages/xchacha20poly1305/xchacha20poly1305.ts:
```diff
@@ -0,0 +1,169 @@
+     * decrypt by calling into chacha20poly1305 as is described should be done
+     * in draft-
+     */
+    chachaPoly1305.open(modifiedNonce, sealed, associatedData, result);
```
`open` returns `null` if authentication fails. Need to check for that and also return `null` otherwise, the result will be equal to destination or new allocated array instead of null:

```
if (!chachaPoly1305.open(...)) {
    return null;
}
```

But since we also can leave it to chachaPoly1305 to deal with allocating destination, we can just do:

```
const chachaPoly1305 = new ChaCha20Poly1305(subKey);
const result = chachaPoly1305.open(modifiedNonce, sealed, associatedData, dst);
wipe(modifiedNonce);
return result;
```

kdenhartog replied:
I was curious what your opinion was on this. I found it somewhat difficult to figure out exactly how I should be calling into `XChacha20` and generating the tag because the RFC defines the usage of `AEAD_Chacha20_poly1305`. Do you think it would be more useful to call directly to `XChacha20` in this function rather than call to `Chacha20poly1305`?

dchest commented on yarn.lock:
```diff
@@ -2,12 +2,843 @@
 # yarn lockfile v1
```
Lock files are infeasible to review, but can introduce *anything* into the repo :), so I'll regenerate it myself from the fresh package.json.

kdenhartog replied:
Definitely the safe bet, given some of the upstream attacks that have occurred in the past.

dchest commented on package.json:
```diff
@@ -22,7 +22,7 @@
     "browserify": "^14.4.0",
     "jasmine": "^2.6.0",
-    "lerna": "^2.0.0-rc.5",
+    "lerna": "^3.16.4",
```
I had some problems publishing using recent versions, so I left it `rc.5`. Maybe the are fixed, but I'd prefer to update it myself; and maybe even switch to yarn-native multipackage support.

kdenhartog replied:
The reason I switched to this is that I was having difficulties getting the dependencies to find each other when I was doing `yarn build` and this was the only solution I found for it.

kdenhartog commented on packages/xchacha20/xchacha20.ts:
```diff
@@ -0,0 +1,224 @@
+
+    // Use last 8 bytes of 24-byte extended nonce as an actual nonce prefixed by 4 NUL bytes,
+    // and a subkey derived in the previous step as key to encrypt.
+    const modified_nonce = new Uint8Array(12);
```
fixed, thanks for pointing this one out.

kdenhartog commented on packages/xchacha20/xchacha20.ts:
```diff
@@ -0,0 +1,224 @@
+        throw new Error("XChacha20 nonce must be 24 bytes");
+    }
+
+    // Use HSalsa one-way function to transform first 16 bytes of
```
fixed.

kdenhartog commented:
Thanks for the review. I also noticed that for some reason the CI pipeline was failing to import `ChaCha20Poly1305` into the `XChaCha20Poly1305 module`. I wonder if it's because of the modifications I made around lerna.

kdenhartog commented:
I was just poking around this branch and realized the benchmarking wasn't properly fixed. I'm going to be submitting a commit for that soon. Please don't merge before I remove the "WIP" tag on the PR title.

kdenhartog commented:
Everything should be working now. I had noticed that the benchmarking was broken when I was seeing the speed of operations and fixed it. I've squashed all the commits for the comments and the fixes for benchmarking together into commit `b96e05e`

dchest commented:
Thank you, merged! I'll tweak the build process and then release it.
